### PR TITLE
Improvements to foreach

### PR DESCRIPTION
**Pull request by tpluscode**
* updated code to reflect recent changes from #15
* added test which runs the example
* extended the stream so that it can also run unconnected stream (pipeline starting with `Readable`)

### CODE REVIEW

File: lib/forEach.js [hunk 1]
```diff
@@ -15,36 +17,45 @@ class ObjectToReadable extends Readable {
...
-  static create (node) {
-    return new ForEach(node, this.pipeline)
+  runPipeline (chunk, pipeline) {
+    if (this.handleChunk) {
+      this.handleChunk.call(pipeline, chunk)
+    }
+
+    return pipeline.init().then(() => {
+      if (!isWritable(pipeline.streams[0])) {
+        return run(pipeline.streams[0])
```
Comment (tpluscode) [inline lib/forEach.js:49]: if first stream is no writable then we cannot `.pipe()`

File: lib/forEach.js [hunk 2]
```diff
@@ -15,36 +17,45 @@ class ObjectToReadable extends Readable {
...
     current.on('data', chunk => this.push(chunk))
-    current.on('end', () => callback())
     current.on('error', err => callback(err))
   }
 
-  static create (node) {
-    return new ForEach(node, this.pipeline)
+  runPipeline (chunk, pipeline) {
+    if (this.handleChunk) {
+      this.handleChunk.call(pipeline, chunk)
```
Comment (tpluscode) [inline lib/forEach.js:44]: This allows the caller to store the chunk as variable for example

**Review by tpluscode — COMMENTED** [review]
@bergos please have a look

File: lib/forEach.js (hunk 2 above)
Comment (tpluscode) [reply lib/forEach.js:44]: Here's an example usage

```
<#doTransform> a :Step ;
   :operation [
       code:link <node:barnard59-core#forEach> ;
       a code:ecmaScript
     ] ;
   :arguments (
     <#TransformCsv>
     "fileName => this.variables.set('csv', fileName)"^^code:ecmaScript
   ) .
```

**Comment by bergos** [comment]
Is it be possible to use the loader to load the pipeline? That way the parts of the sequence can be whatever as long as there is a loader which returns a stream.

Should I merge it?

**Comment by tpluscode** [comment]
Not sure what you're asking. The pipelines are already loaded with as loader.

**Comment by bergos** [comment]
Sorry, mixed it up with the sequence code. Maybe we could use the loader [here](https://github.com/zazuko/barnard59-core/pull/18/files#diff-2bc260a7e5232288fe07948189b63be9R29), but think if we do it, it should be done in a second step.

**Comment by tpluscode** [comment]
> That way the parts of the sequence can be whatever as long as there is a loader which returns a stream.

About that, I think it would duplicate what we already have with pipeline steps. If we decide to bring sequence into the core lib (or `barnard59-base`) then I'd prefer to keep it focused on chaining pipelines. Otherwise it's going to create two way to achieve too similar goals.

Technically though the arguments of a `sequence` or `forEach` are indeed loader using the registry already so it's wouldn't require much changes to have them execute any stream and not only pipelines.

> Maybe we could use the loader here

One obvious difference is that the `clone` call is more powerful and enforces the `objectMode: true` switch. The loader does not allow that ATM.

We could allow that by adding this parameter to the loader, though that will also add changes to the registry itself.

```diff
-function loader (term, dataset, { context, variables, basePath, loaderRegistry }) {
+function loader (term, dataset, { context, variables, basePath, loaderRegistry }, objectMode) {
```

Will it not conflict conceptually with the loader's logic of setting `objectMode` base on the pipeline's declared `rdf:type`?

**Pull request merged**